lib/libc/sys/swapctl.2
.\"	$NetBSD: swapctl.2,v 1.14 2000/01/13 12:11:30 ad Exp $
.\"
.\" Copyright (c) 1997 Matthew R. Green
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd June 20, 1997
.Dt SWAPCTL 2
.Os
.Sh NAME
.Nm swapctl
.Nd modify swap configuration
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.Fd #include <unistd.h>
.Fd #include <sys/swap.h>
.Ft int
.Fn swapctl "int cmd" "const void *arg" "int misc"
.Sh DESCRIPTION
The
.Nm
function is used to add and delete swap devices, and modify their
configuration.
.Pp
The
.Fa cmd
parameter specifies the operation to be performed.  The
.Fa arg
and
.Fa misc
parameters have different meanings, depending on the
.Fa cmd
parameter.
.Bl -item -offset indent
.It
If
.Fa cmd
is
.Dv SWAP_NSWAP ,
the current number of swap devices in the system is returned.  The
.Fa arg
and
.Fa misc
parameters are ignored.
.It
If
.Fa cmd
is
.Dv SWAP_STATS ,
the current statistics for swap devices are returned in the
.Fa arg
parameter.  No more than
.Fa misc
swap devices are returned.  The
.Fa arg
parameter should point to an array of at least
.Fa misc
struct swapent structures:
.Bd -literal
struct swapent {
	dev_t	se_dev;			/* device id */
	int	se_flags;		/* entry flags */
	int	se_nblks;		/* total blocks */
	int	se_inuse;		/* blocks in use */
	int	se_priority;		/* priority */
	char	se_path[PATH_MAX+1];	/* path to entry */
};
.Ed
.Pp
The flags are defined as
.Bd -literal
	SWF_INUSE       in use: we have swapped here
	SWF_ENABLE      enabled: we can swap here
	SWF_BUSY        busy: I/O happening here
	SWF_FAKE        fake: still being built
.Ed
.It
If
.Fa cmd
is
.Dv SWAP_ON ,
the
.Fa arg
parameter is used as a pathname of a file to enable swapping to.
The
.Fa misc
parameter is used to set the priority of this swap device.
.It
If
.Fa cmd
is
.Dv SWAP_OFF ,
the
.Fa arg
parameter is used as the pathname of a file to disable swapping from.
The
.Fa misc
parameter is ignored.
.It
If
.Fa cmd
is
.Dv SWAP_CTL ,
the
.Fa arg
and
.Fa misc
parameters have the same function as for the
.Dv SWAP_ON
case, except that they change the priority of a currently enabled swap device.
.El
.Pp
When swapping is enabled on a block device, the first portion of the disk is
left unused to prevent any disklable present from being overwritten.  This
space is allocated from the swap device when the
.Dv SWAP_ON
command is used.
.Sh RETURN VALUES
If the
.Fa cmd
parameter is
.Dv SWAP_NSWAP
or
.Dv SWAP_STATS ,
.Fn swapctl
returns the number of swap devices, if successful.  The
.Dv SWAP_NSWAP
command is always successful.  Otherwise it returns 0 on success and -1
on failure, setting the global variable
.Va errno
to indicate the error.
.Sh ERRORS
.Fn swapctl
succeeds unless:
.Bl -tag -width Er
.It Bq Er ENOTDIR
A component of the path prefix is not a directory.
.It Bq Er ENAMETOOLONG
A component of a pathname exceeded
.Dv NAME_MAX
characters, or an entire path name exceeded
.Dv PATH_MAX
characters.
.It Bq Er ENOENT
The named device does not exist.  For the
.Dv SWAP_CTL
command, the named device is not currently enabled for swapping.
.It Bq Er EACCES
Search permission is denied for a component of the path prefix.
.It Bq Er ELOOP
Too many symbolic links were encountered in translating the pathname.
.It Bq Er EPERM
The caller is not the super-user.
.It Bq Er EBUSY
The device specified by
.Fa arg
has already been made available for swapping.
.It Bq Er EINVAL
The device configured by
.Fa arg
has no associated size, or the
.Fa cmd
was unknown.
.It Bq Er ENXIO
The major device number of
.Fa arg
is out of range (this indicates no device driver exists
for the associated hardware).
.It Bq Er EIO
An I/O error occurred while opening the swap device.
.It Bq Er EFAULT
.Fa arg
points outside the process' allocated address space.
.El
.Sh SEE ALSO
.Xr swapctl 8 ,
.Xr config 8
.Sh AUTHOR
The current swap system was designed and implemented by Matthew Green
<mrg@eterna.com.au>, with help from Paul Kranenburg <pk@NetBSD.ORG>
and Leo Weppelman <leo@NetBSD.ORG>, and insights from Jason R. Thorpe
<thorpej@NetBSD.ORG>.
.Sh HISTORY
The
.Fn swapctl
function call appeared in
.Nx 1.3 .
The
.Fa se_path
member was added to
.Va struct swapent
in
.Nx 1.4 ,
when the header file was also moved from
.Pa <vm/vm_swap.h> .


sbin/swapctl/swapctl.c
/*	$NetBSD: swapctl.c,v 1.11 1999/04/26 01:02:25 mrg Exp $	*/

/*
 * Copyright (c) 1996, 1997, 1999 Matthew R. Green
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * swapctl command:
 *	-A		add all devices listed as `sw' in /etc/fstab (also
 *			(sets the the dump device, if listed in fstab)
 *	-D <dev>	set dumpdev to <dev>
 *	-t [blk|noblk]	if -A, add either all block device or all non-block
 *			devices
 *	-a <dev>	add this device
 *	-d <dev>	remove this swap device (not supported yet)
 *	-l		list swap devices
 *	-s		short listing of swap devices
 *	-k		use kilobytes
 *	-p <pri>	use this priority
 *	-c		change priority
 *
 * or, if invoked as "swapon" (compatibility mode):
 *
 *	-a		all devices listed as `sw' in /etc/fstab
 *	-t		same as -t above (feature not present in old
 *			swapon(8) command)
 *	<dev>		add this device
 */

#include <sys/param.h>
#include <sys/stat.h>

#include <vm/vm_swap.h>

#include <unistd.h>
#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fstab.h>

#include "swapctl.h"

int	command;

/*
 * Commands for swapctl(8).  These are mutually exclusive.
 */
#define	CMD_A		0x01	/* process /etc/fstab */
#define	CMD_D		0x02	/* set dumpdev */
#define	CMD_a		0x04	/* add a swap file/device */
#define	CMD_c		0x08	/* change priority of a swap file/device */
#define	CMD_d		0x10	/* delete a swap file/device */
#define	CMD_l		0x20	/* list swap files/devices */
#define	CMD_s		0x40	/* summary of swap files/devices */

#define	SET_COMMAND(cmd) \
do { \
	if (command) \
		usage(); \
	command = (cmd); \
} while (0)

/*
 * Commands that require a "path" argument at the end of the command
 * line, and the ones which require that none exist.
 */
#define	REQUIRE_PATH	(CMD_D | CMD_a | CMD_c | CMD_d)
#define	REQUIRE_NOPATH	(CMD_A | CMD_l | CMD_s)

/*
 * Option flags, and the commands with which they are valid.
 */
int	kflag;		/* display in 1K blocks */
#define	KFLAG_CMDS	(CMD_l | CMD_s)

int	pflag;		/* priority was specified */
#define	PFLAG_CMDS	(CMD_A | CMD_a | CMD_c)

char	*tflag;		/* swap device type (blk or noblk) */
#define	TFLAG_CMDS	(CMD_A)

int	pri;		/* uses 0 as default pri */

static	void change_priority __P((const char *));
static	void add_swap __P((const char *));
static	void delete_swap __P((const char *));
static	void set_dumpdev __P((const char *));
	int  main __P((int, char *[]));
static	void do_fstab __P((void));
static	void usage __P((void));
static	void swapon_command __P((int, char **));
#if 0
static	void swapoff_command __P((int, char **));
#endif

extern	char *__progname;	/* from crt0.o */

int
main(argc, argv)
	int	argc;
	char	*argv[];
{
	int	c;

	if (strcmp(__progname, "swapon") == 0) {
		swapon_command(argc, argv);
		/* NOTREACHED */
	}

#if 0
	if (strcmp(__progname, "swapoff") == 0) {
		swapoff_command(argc, argv);
		/* NOTREACHED */
	}
#endif

	while ((c = getopt(argc, argv, "ADacdlkp:st:")) != -1) {
		switch (c) {
		case 'A':
			SET_COMMAND(CMD_A);
			break;

		case 'D':
			SET_COMMAND(CMD_D);
			break;

		case 'a':
			SET_COMMAND(CMD_a);
			break;

		case 'c':
			SET_COMMAND(CMD_c);
			break;

		case 'd':
			SET_COMMAND(CMD_d);
			break;

		case 'l':
			SET_COMMAND(CMD_l);
			break;

		case 'k':
			kflag = 1;
			break;

		case 'p':
			pflag = 1;
			/* XXX strtol() */
			pri = atoi(optarg);
			break;

		case 's':
			SET_COMMAND(CMD_s);
			break;

		case 't':
			if (tflag != NULL)
				usage();
			tflag = optarg;
			break;

		default:
			usage();
			/* NOTREACHED */
		}
	}

	/* Did the user specify a command? */
	if (command == 0)
		usage();

	argv += optind;
	argc -= optind;

	switch (argc) {
	case 0:
		if (command & REQUIRE_PATH)
			usage();
		break;

	case 1:
		if (command & REQUIRE_NOPATH)
			usage();
		break;

	default:
		usage();
	}

	/* To change priority, you have to specify one. */
	if ((command == CMD_c) && pflag == 0)
		usage();

	/* Sanity-check -t */
	if (tflag != NULL) {
		if (command != CMD_A)
			usage();
		if (strcmp(tflag, "blk") != 0 &&
		    strcmp(tflag, "noblk") != 0)
			usage();
	}

	/* Dispatch the command. */
	switch (command) {
	case CMD_l:
		list_swap(pri, kflag, pflag, 0, 1);
		break;

	case CMD_s:
		list_swap(pri, kflag, pflag, 0, 0);
		break;

	case CMD_c:
		change_priority(argv[0]);
		break;

	case CMD_a:
		add_swap(argv[0]);
		break;

	case CMD_d:
		delete_swap(argv[0]);
		break;

	case CMD_A:
		do_fstab();
		break;

	case CMD_D:
		set_dumpdev(argv[0]);
		break;
	}

	exit(0);
}

/*
 * swapon_command: emulate the old swapon(8) program.
 */
void
swapon_command(argc, argv)
	int argc;
	char **argv;
{
	int ch, fiztab = 0;

	while ((ch = getopt(argc, argv, "at:")) != -1) {
		switch (ch) {
		case 'a':
			fiztab = 1;
			break;
		case 't':
			if (tflag != NULL)
				usage();
			tflag = optarg;
			break;
		default:
			goto swapon_usage;
		}
	}
	argc -= optind;
	argv += optind;

	if (fiztab) {
		if (argc)
			goto swapon_usage;
		/* Sanity-check -t */
		if (tflag != NULL) {
			if (strcmp(tflag, "blk") != 0 &&
			    strcmp(tflag, "noblk") != 0)
				usage();
		}
		do_fstab();
		exit(0);
	} else if (argc == 0 || tflag != NULL)
		goto swapon_usage;

	while (argc) {
		add_swap(argv[0]);
		argc--;
		argv++;
	}
	exit(0);
	/* NOTREACHED */

 swapon_usage:
	fprintf(stderr, "usage: %s -a [-t blk|noblk]\n", __progname);
	fprintf(stderr, "       %s <path> ...\n", __progname);
	exit(1);
}

/*
 * change_priority:  change the priority of a swap device.
 */
void
change_priority(path)
	const char	*path;
{

	if (swapctl(SWAP_CTL, path, pri) < 0)
		warn("%s", path);
}

/*
 * add_swap:  add the pathname to the list of swap devices.
 */
void
add_swap(path)
	const char *path;
{
	struct stat sb;

	if (stat(path, &sb) < 0)
		goto oops;

	if (sb.st_mode & S_IROTH)
		warnx("%s is readable by the world", path);
	if (sb.st_mode & S_IWOTH)
		warnx("%s is writable by the world", path);

	if (swapctl(SWAP_ON, path, pri) < 0)
oops:
		err(1, "%s", path);
}

/*
 * delete_swap:  remove the pathname to the list of swap devices.
 */
void
delete_swap(path)
	const char *path;
{

	if (swapctl(SWAP_OFF, path, pri) < 0)
		err(1, "%s", path);
}

void
set_dumpdev(path)
	const char *path;
{

	if (swapctl(SWAP_DUMPDEV, path, NULL) == -1)
		warn("could not set dump device to %s", path);
	else
		printf("%s: setting dump device to %s\n", __progname, path);
}

void
do_fstab()
{
	struct	fstab *fp;
	char	*s;
	long	priority;
	struct	stat st;
	int	isblk;
	int	gotone = 0;

#define PRIORITYEQ	"priority="
#define NFSMNTPT	"nfsmntpt="
#define PATH_MOUNT	"/sbin/mount_nfs"
	while ((fp = getfsent()) != NULL) {
		const char *spec;

		spec = fp->fs_spec;

		if (strcmp(fp->fs_type, "dp") == 0) {
			set_dumpdev(spec);
			continue;
		}

		if (strcmp(fp->fs_type, "sw") != 0)
			continue;
		isblk = 0;

		if ((s = strstr(fp->fs_mntops, PRIORITYEQ)) != NULL) {
			s += sizeof(PRIORITYEQ) - 1;
			priority = atol(s);
		} else
			priority = pri;

		if ((s = strstr(fp->fs_mntops, NFSMNTPT)) != NULL) {
			char *t, cmd[2*PATH_MAX+sizeof(PATH_MOUNT)+2];

			/*
			 * Skip this song and dance if we're only
			 * doing block devices.
			 */
			if (tflag != NULL &&
			    strcmp(tflag, "blk") == 0)
				continue;

			t = strpbrk(s, ",");
			if (t != 0)
				*t = '\0';
			spec = strdup(s + strlen(NFSMNTPT));
			if (t != 0)
				*t = ',';

			if (spec == NULL)
				errx(1, "Out of memory");

			if (strlen(spec) == 0) {
				warnx("empty mountpoint");
				free((char *)spec);
				continue;
			}
			snprintf(cmd, sizeof(cmd), "%s %s %s",
				PATH_MOUNT, fp->fs_spec, spec);
			if (system(cmd) != 0) {
				warnx("%s: mount failed", fp->fs_spec);
				continue;
			}
		} else {
			/*
			 * Determine blk-ness.
			 */
			if (stat(spec, &st) < 0) {
				warn(spec);
				continue;
			}
			if (S_ISBLK(st.st_mode))
				isblk = 1;
		}

		/*
		 * Skip this type if we're told to.
		 */
		if (tflag != NULL) {
			if (strcmp(tflag, "blk") == 0 && isblk == 0)
				continue;
			if (strcmp(tflag, "noblk") == 0 && isblk == 1)
				continue;
		}

		if (swapctl(SWAP_ON, spec, (int)priority) < 0)
			warn("%s", spec);
		else {
			gotone = 1;
			printf("%s: adding %s as swap device at priority %d\n",
			    __progname, fp->fs_spec, (int)priority);
		}

		if (spec != fp->fs_spec)
			free((char *)spec);
	}
	if (gotone == 0)
		exit(1);
}

void
usage()
{

	fprintf(stderr, "usage: %s -A [-p priority] [-t blk|noblk]\n",
	    __progname);
	fprintf(stderr, "       %s -D dumppath\n", __progname);
	fprintf(stderr, "       %s -a [-p priority] path\n", __progname);
	fprintf(stderr, "       %s -c -p priority path\n", __progname);
	fprintf(stderr, "       %s -d path\n", __progname);
	fprintf(stderr, "       %s -l | -s [-k]\n", __progname);
	exit(1);
}
